examples/jenkins-image-sample.groovy

```diff
@@ -0,0 +1,119 @@
+    def saSelector4 = openshift.selector( "dc", [ "template" : "mytemplate" ] )
+    
+    def templateSelector = openshift.selector( "template", "mongodb-ephemeral")
+        
+    templateSelector.delete('--ignore-not-found')
+    openshift.delete('all,secret', '-l', 'template=mongodb-ephemeral-template', '--ignore-not-found')
+    template = openshift.create('https://raw.githubusercontent.com/openshift/origin/master/examples/db-templates/mongodb-ephemeral-template.json').object()
+    
+    // Explore the Groovy object which models the OpenShift template as a Map
+    echo "Template contains ${template.parameters.size()} parameters"
+
+    // For fun, modify the template easily while modeled in Groovy
+    template.labels["mylabel"] = "myvalue"
+
+    // Process the modeled template. We could also pass JSON/YAML or a template name instead.
+    // note: -p option for oc process not in the oc version that we currently ship with openshift jenkins images
+    def objectModels = openshift.process( template )//, "-p", "MEMORY_LIMIT=600Mi")
+
+    // objectModels is a list of objects the template defined, modeled as Groovy objects
+    echo "The template will create ${objectModels.size()} objects"
+
+    // For fun, modify the objects that have been defined by processing the template
+    for ( o in objectModels ) {
+        o.metadata.labels[ "anotherlabel" ] = "anothervalue"
+    }
+
+    // Serialize the objects and pass them to the create API.
+    // We could also pass JSON/YAML directly; openshift.create(readFile('some.json'))
+    def created = openshift.create(objectModels)
+
+    // Create returns a selector which will always select the objects created
+    created.withEach {
+        // Each loop binds the variable 'it' to a selector which selects a single object
+        echo "Created ${it.name()} from template with labels ${it.object().metadata.labels}"
+    }
+
+    // Filter created objects and create a selector which selects only the new DeploymentConfigs
+    def dcs = created.narrow("dc")
+    echo "Database will run in deployment config: ${dcs.name()}"
+    timeout(5) {
+        // Find a least one pod related to the DeploymentConfig and wait it satisfies a condition
+        dcs.related('pods').untilEach(1) {
+            // untilEach only terminates when each selected item causes the body to return true
+            return it.object().status.phase != 'Pending'
+        }
+    }
+
+    // Print out all pods created by the DC
+    echo "Template created pods: ${dcs.related('pods').names()}"
+
+    // Show how we can use labels to select as well
+    echo "Finding dc using labels instead: ${openshift.selector('dc',[mylabel:'myvalue']).names()}"
+
+    echo "DeploymentConfig description"
+    dcs.describe()
+    echo "DeploymentConfig history"
+    dcs.rollout().history()
+
+    def rubySelector = openshift.selector("bc", "ruby")
+    def builds
+    try {
+        rubySelector.object()
```

Comment:
Yeah i doesn't work, but I misinterpreted the hang.

Even though there are no bc's yet in my project (confirmed via use of `oc` from where I ran `oc cluster up`, `oc new-app jenkins-ephemeral`, etc.), `rubySelector.count()` is returning 1.  When I call names to dump the output, it returns `buildconfig/ruby` .

The *hang* stems from the subsequent build watch, as `new-build` was not called and there in fact is not a *ruby* bc yet.  It occurs even if I move the related call after the count call.

I could add debug to confirm, but at a glance, I suspect the `objectList` for the `rubySelector` is being populated on the constructor here:  https://github.com/openshift/jenkins-client-plugin/blob/master/src/main/resources/com/openshift/jenkins/plugins/OpenShiftDSL.groovy#L718

So should I just check for less than or equal to 1 on the count, or should a fix be made to ensure count == 0 when a selector is created ?